gridded/pyugrid/ugrid.py

```diff
@@ -840,44 +840,196 @@ def build_boundaries(self):
+
+    def get_face_edge_orientation(self):
+        """
+        Get the orientation for each edge in the corresponding face
+
+        This method returns an array with the same shape as :attr:`faces` that
+        is one if the corresponding edge has the same orientation as in
+        :attr:`edges`, and -1 otherwise
+        """
+        # we build the face edge connectivity but do not sort the edge nodes.
+        # With this, we will get `num_edges` where the edge is flipped compared
+        # to the definition in :attr:`edges`
+        face_edge_connectivity = self._build_face_edge_connectivity(sort=False)
+        num_edges = self.edges.shape[0]
+        if np.ma.isMA(face_edge_connectivity):
+            return np.ma.where(face_edge_connectivity == num_edges, 1, -1)
+        else:
+            return np.where(face_edge_connectivity == num_edges, 1, -1)
+
+    def _create_dual_edge_mesh(self):
+        """Create a :class:`UGrid` instance that represents the dual edge mesh.
+        """
+        try:
+            from scipy.spatial import cKDTree
+        except ImportError:
+            raise ImportError("The scipy package is required to use "
+                              "UGrid.locatbuild_face_edge_connectivity")
+
+        from .inverse_lookup import lookup
+
+        if self.face_edge_connectivity is None:
+            self.build_face_edge_connectivity()
+
+        face_edge_connectivty = self.face_edge_connectivity
+        orientation = self.get_face_edge_orientation()
+        edges = self.edges
+
+        if np.ma.isMA(face_edge_connectivty):
+            face_edge_connectivty = face_edge_connectivty.filled(len(edges))
+
+        # use -999 as fill value (necessary for edges at the domain boundary)
+        dual_face_node_connectivity = np.ones((len(edges), 4)) * -999
+        dual_face_node_connectivity[:, 0] = edges[:, 0]
+        dual_face_node_connectivity[:, 2] = edges[:, 1]
+
+        # get the first index for the face center nodes
+        if self.face_coordinates is None:
+            self.build_face_coordinates()
+        dual_nodes = np.r_[self.nodes, self.face_coordinates]
+        face_node_start = len(self.nodes)
+
+        # now we have to reverse the face_edge_connectivity and get the
+        # *edge_face_connectivity* (i.e. for every edge, get the (up to) two
+        # adjacent faces
+        edge_idx = np.arange(
+            edges.shape[0], dtype=face_edge_connectivty.dtype
+        )
+
+        edge_face_connectivity = np.asarray(lookup(
+            face_edge_connectivty.ravel().astype(np.int32),
+            edge_idx.astype(np.int32), k=2, fill_value=len(edge_idx),
+        ))
+
+        # now get the orientation for each edge from the `orientation` array
+        mask = edge_face_connectivity < face_edge_connectivty.size
+        edge_orientation = np.ones_like(edge_face_connectivity) * -999
+        edge_orientation[mask] = orientation.ravel()[
+            edge_face_connectivity[mask]
+        ]
```

Comment:
hey @Huite! I am struggling a bit with transforming this part now that I want to use your method:

```python
        mask = edge_face_connectivity < face_edge_connectivty.size
        edge_orientation = np.ones_like(edge_face_connectivity) * -999
        edge_orientation[mask] = orientation.ravel()[
            edge_face_connectivity[mask]
        ]
```

Any ideas? `orientation` has the same shape as the `face_edge_connectivity` and stores the orientation for the edge at the same location in `face_edge_connectivity`. We'd also need to transform this into an array with the same shape as the `edge_face_connectivity`.

Previously, this was not so much of an issue, because the `edge_face_connectivity` in my first attempt referred to the position in the raveled `face_edge_connectivty`

```python
        edge_face_connectivity = np.asarray(lookup(
            face_edge_connectivty.ravel().astype(np.int32),
            edge_idx.astype(np.int32), k=2, fill_value=len(edge_idx),
        ))
```
but this is not the case anymore when using your method.

Reply:
alright, I just added the necessary lines to the method `build_edge_face_connectivity`. so this now builds the `edge_face_connectivity` and the `edge_orientation`.